sympy/matrices/matrices.py

```diff
@@ -783,6 +784,23 @@ def __repr__(self):
         return "MatrixKind(%s)" % self.element_kind
 
 
+@Mul._kind_dispatcher.register(MatrixKind, type(NumberKind))
```

Comment:
Why do we need `type(NumberKind)`? Can the type not be referenced directly?

I see we have
```
NumberKind = NumberKind()
```
Maybe it would be better to give the class a different name from the instance. Or maybe the class itself should be used as the Kind object... I think we agreed that we would need that to use Kinds as generic types.

Reply:
Will `_NumberKind` for class name be suffice?

Reply:
Yes.

Side note: is there any reason to even have an instance? Could the class NumberKind itself not serve as the kind object?

Reply:
Changing `NumberKind` that way is not a problem, but I can't imagine how we should implement `MatrixKind(NumberKind)` with that approach.

Reply:
I'm always wary of metaclasses so I don't know if I actually think that this is a good idea. With that caveat maybe something like this:
```python
class MatrixKindMeta(type):
    def __getitem__(self, kind):
        name = f'MatrixKind[{kind.__name__}]'
        return type(name, (self,), {'element_kind': kind})

class MatrixKind(metaclass=MatrixKindMeta):
    element_kind = None
```
Then you have:
```python
In [25]: MatrixKind
Out[25]: MatrixKind

In [26]: MatrixKind[int]
Out[26]: MatrixKind[int]

In [27]: issubclass(MatrixKind[int], MatrixKind)
Out[27]: True
```
Since there is a subclass relationship that would make it possible to do multiple dispatch or type hinting for either `MatrixKind` or `MatrixKind[int]`. I'm not sure how it would interact with the generic types idea.

Reply:
Does this dynamically generate a class every time `MatrixKind[int]` is encountered?

Reply:
Ping @Upabjojr @eric-wieser . Any idea to combine generic with kind dispatcher?

Reply:
Yes it does. There should be a mechanism for returning the same class each time though.

Reply:
Will this new design seriously affect the way how kind is dispatched in this PR? If not, I would like to introduce kind dispatching first (this PR), remove postprocessor with next PR, and then attend to this matter.

Reply:
Fair enough. It was just an idea for consideration. I see now that there is potentially a need to multiple dispatch with kinds which means that it maybe makes sense for all kinds to be classes rather than instances.